File: dotnet/DesignPatterns/StructuralDesignPatterns/CompositeDesignPattern/CompositeDesignPattern.Example1/Box.cs
namespace CompositeDesignPattern.Example1;

public class Box : OrderItem
{
	private readonly List<OrderItem> _items = new();
	private readonly double _boxWeight;

	public Box(string name, double boxWeight = 100) : base(name)
	{
		_boxWeight = boxWeight;
	}

	public override double GetWeight()
	{
		return _boxWeight + _items.Sum(i => i.GetWeight());
	}

	public void AddItem(OrderItem item)
	{
		_items.Add(item);
	}

	public void RemoveItem(OrderItem item)
	{
		_items.Remove(item);
	}
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/CompositeDesignPattern/CompositeDesignPattern.Example1/src/Box.ts
import { OrderItem } from './OrderItem';

export class Box extends OrderItem {
  private readonly items: OrderItem[] = [];

  constructor(name: string, private readonly boxWeight: number = 100) { super(name); }

  getWeight(): number {
    return this.boxWeight + this.items.reduce((sum, i) => sum + i.getWeight(), 0);
  }

  addItem(item: OrderItem): void { this.items.push(item); }
  removeItem(item: OrderItem): void {
    const idx = this.items.indexOf(item);
    if (idx !== -1) this.items.splice(idx, 1);
  }
}
